### PR TITLE
Implement first stage gpu optimizer

### PR DESCRIPTION
**Pull request by ttochi**
### Feature
In staged optimizer, parameter update (cpu job) is overlapped by fwd, bwd computation (gpu job).
However, GPU do nothing while paramter update of first stage.
This PR resolve this by apply gpu optimizer for first stage and this can be run by add `--spiral-heterogeneous-optimizer` option.

- Add `SpiralGPUAdam` which is based on `FusedAdam` of NVIDIA apex.
- Modify schedule logic to support gpu_optimizer stage.
- (misc) Keep `DeepSpeedCPUAdam` as baseline and create `DeepSpeedFloat16Optimizer` which is divided from `SpiralFloat16Optimizer`

### TODO
- I will move some mixed precision logic in `SpiralFloat16Optimizer` to `SpiralGPUOptimizer`
- Currently, first stage gpu optimizer hold optimizer state (momentum, variance, param32). To save gpu memory, we need to prefetch and offload optimizer state per parameter tensor.

(+) Reslove #15 and #30

### CODE REVIEW

File: megatron/arguments.py [hunk 1]
```diff
@@ -439,6 +439,9 @@ def validate_args(args, defaults={}):
         if args.use_distributed_optimizer:
             raise RuntimeError(
                 "SpiralPipe currently does not support distributed optimizer")
+        if args.spiral_heterogeneous_optimizer:
+            assert args.spiral_stage_optimizer, "spiral-stage-optimizer should be enabled with spiral-heterogeneous-optimizer"
```
Comment (gajagajago) [inline megatron/arguments.py]:
```suggestion
            assert args.spiral_stage_optimizer, "spiral-heterogeneous-optimizer should be enabled with spiral-stage-optimizer"
```

File: megatron/spiral/schedules/mobius_schedule.py [hunk 2]
```diff
@@ -526,7 +523,8 @@ def _wait_reqs(reqs: List[Work]):
                 raise RuntimeError("wait_event failed")
 
             # free bwd stage
-            model[bwd_stage_id].spiral_free()
+            if not optimize_after_bwd_stage or optimizer.is_cpu_optimizer(bwd_stage_id):
+                model[bwd_stage_id].spiral_free()
             free_curr = get_thunder_cuda_manager().Event(
```
Comment (gajagajago) [inline megatron/spiral/schedules/mobius_schedule.py:535]: Why is 528-535 not under 526 if block?

File: megatron/spiral/schedules/mobius_schedule.py [hunk 3]
```diff
@@ -545,16 +543,19 @@ def _wait_reqs(reqs: List[Work]):
...
                 if optimize_after_bwd_stage:
                     optimizer[bwd_stage_id].reduce_model_grads(get_args(), get_timers())
                 else:
                     model[bwd_stage_id].allreduce_gradients()
+
                 # offload grads
-                model[bwd_stage_id].spiral_offload_grad(non_blocking=True)
+                if not optimize_after_bwd_stage or optimizer.is_cpu_optimizer(bwd_stage_id):
+                    model[bwd_stage_id].spiral_offload_grad(non_blocking=True)
                 offload_grad_curr = get_thunder_cuda_manager().Event(
```
Comment (gajagajago) [inline megatron/spiral/schedules/mobius_schedule.py:563]: Why is 556-563 not under 554 if block?

File: megatron/spiral/schedules/mobius_schedule.py [hunk 4]
```diff
@@ -526,7 +523,8 @@ def _wait_reqs(reqs: List[Work]):
                 raise RuntimeError("wait_event failed")
 
             # free bwd stage
-            model[bwd_stage_id].spiral_free()
+            if not optimize_after_bwd_stage or optimizer.is_cpu_optimizer(bwd_stage_id):
```
Comment (gajagajago) [inline megatron/spiral/schedules/mobius_schedule.py]: 이 `or`의 앞부분 넣은거 너무 헷갈림.. schedule에서 다 이렇게 분기처리 한거 같은데 (e.g., 554) 다른 방법 제시좀..

File: megatron/spiral/schedules/spipe_schedule.py [hunk 5]
```diff
@@ -626,11 +626,25 @@ def _wait_reqs(reqs: List[Work]):
             # flush free grad event queries
```
Comment (gajagajago) [inline megatron/spiral/schedules/spipe_schedule.py]:
```suggestion
            # flush offload grad event queries
```

File: megatron/spiral/schedules/spipe_schedule.py [hunk 6]
```diff
@@ -626,11 +626,25 @@ def _wait_reqs(reqs: List[Work]):
...
+                    sync=True
                 )
                 == -1
             ):
                 raise RuntimeError("wait_event failed")
 
+    if optimize_after_bwd_stage:
+        for bwd_stage_id in range(mpu.get_spiral_backward_virtual_size() - 1, -1, -1):
+            if not optimizer.is_cpu_optimizer(bwd_stage_id):
+                # flush free grad event queries
```
Comment (gajagajago) [inline megatron/spiral/schedules/spipe_schedule.py]:
```suggestion
                # flush offload param event queries
```

Comment (gajagajago) [reply megatron/spiral/schedules/spipe_schedule.py]: + 596 if문 내에서 둘다 처리하는게 보기편할듯

File: megatron/spiral/schedules/spipe_schedule.py
Comment (gajagajago) [inline megatron/spiral/schedules/spipe_schedule.py:1]: ditto mobius_schedule.py

File: megatron/optimizer/__init__.py [hunk 7]
```diff
@@ -62,7 +63,8 @@ def get_param_groups(modules,
                 no_wd = no_weight_decay_cond(name, param)
             else:
                 # do not regularize biases nor Norm parameters
-                no_wd = name.endswith(".bias") or len(param.shape) == 1
+                shape = param.spiral_shape if args.spiral and args.spiral_stage_optimizer else param.shape
```
Comment (gajagajago) [inline megatron/optimizer/__init__.py:66]: I don't understand this.

File: megatron/optimizer/__init__.py [hunk 8]
```diff
@@ -146,13 +144,24 @@ def get_megatron_optimizer(model,
...
+
+            # NOTE (SpiralPipe) Spiral stage optimizer uses SpiralCPUAdam, which overlaps weight update with upstream bwd stage computation.
+            # If --spiral-heterogeneous-optimizer is enabled, apply gpu optimizer to first stage.
+            if args.spiral_heterogeneous_optimizer and _unwrapped_model.spiral_backward_stage_id == 0:
+                from megatron.spiral.optimizer.gpu_adam import SpiralGPUAdam
+                inner_opt_ty = SpiralGPUAdam
+            else:
+                from megatron.spiral.optimizer.cpu_adam import SpiralCPUAdam
+                inner_opt_ty = SpiralCPUAdam
+                params_have_main_grad = False
```
Comment (gajagajago) [inline megatron/optimizer/__init__.py:163]: Please preserve my original NOTE on this.
```
# NOTE (SpiralPipe) `params` here annotates the "optimizer params". It is not always the same as the params referenced during training. For SpiralPipe, the optimizer params are the offloaded params and hence should only have grad field, while the params referred during can have both main_grad and grad field. Hence, setting `params_have_main_grad` to True incurs explicit copy from main_grad into grad (optimizer.step() calls it), which is not necessary.
```

File: megatron/optimizer/__init__.py [hunk 9]
```diff
@@ -146,13 +144,24 @@ def get_megatron_optimizer(model,
...
         else:
             from deepspeed.ops.adam import DeepSpeedCPUAdam
             inner_opt_ty = DeepSpeedCPUAdam
+            params_have_main_grad = False
```
Comment (gajagajago) [inline megatron/optimizer/__init__.py:167]: no matter `DDP_impl`??

File: megatron/optimizer/optimizer.py [hunk 10]
```diff
@@ -424,7 +425,9 @@ def step(self, args, timers):
 
             # We are done with scaling gradients
             # so we can update the loss scale.
-            self.grad_scaler.update(found_inf_flag)
+            # TODO: Remove this condition and self.use_global_grad_scaler (dummy flag)
+            if not self.use_global_grad_scaler:
+                self.grad_scaler.update(found_inf_flag)
```
Comment (gajagajago) [inline megatron/optimizer/optimizer.py:430]: what is global grad scaler???

File: megatron/spiral/init_context.py [hunk 11]
```diff
@@ -370,12 +370,12 @@ def _offload_data(param, non_blocking=False):
...
                 ), f"Offload tensor shape mismatch ({param.spiral_tensor.shape} != {param.data.shape})"
                 param.spiral_tensor.copy_(param.data, non_blocking=non_blocking)
-            if not non_blocking:
-                # NOTE: for non-blocking offload, spiral_status should be changed after waiting in the caller
-                param.spiral_status = SpiralParamStatus.CPU
 
         def _fetch_data(param, non_blocking=False):
-            assert param.spiral_status == SpiralParamStatus.CPU
+            # if parameter is already in gpu memory, skip fetch
+            if param.spiral_status == SpiralParamStatus.GPU:
```
Comment (gajagajago) [inline megatron/spiral/init_context.py:377]: Please add some NOTE and warning (using warning.warn) that this should be only true under very specific circumstances. Most cases should naturally fetch from CPU in spiral case.

File: megatron/spiral/init_context.py [hunk 12]
```diff
@@ -469,11 +469,13 @@ def _free_grad(param: Parameter) -> None:
             if params_have_main_grad:
                 assert hasattr(param, "main_grad")
                 # Ensure the tensor memory is not reused for another tensor until all work queued on stream is complete
-                param.main_grad.record_stream(get_thunder_cuda_manager().Stream('offload'))
+                if param.main_grad is not None:
```
Comment (gajagajago) [inline megatron/spiral/init_context.py:473]: why do we need these if statements? (472, 477)

**Review by gajagajago — COMMENTED** [review]
I left reviews

File: megatron/spiral/schedules/mobius_schedule.py (hunk 4 above)
Comment (ttochi) [reply megatron/spiral/schedules/mobius_schedule.py]: 아래 method를 만들어서 처리하는건 어떨지?
```py
def is_cpu_optimizer(bwd_stage_id):
    if get_args().spiral_stage_optimizer:
        return optimizer.is_cpu_optimizer(bwd_stage_id)
    else:
        return True
```

File: megatron/spiral/schedules/mobius_schedule.py (hunk 2 above)
Comment (ttochi) [reply megatron/spiral/schedules/mobius_schedule.py:535]: If we create the event inside if statement, we need to make branch at the event wait logic. It will be complicated...

File: megatron/optimizer/__init__.py (hunk 9 above)
Comment (ttochi) [reply megatron/optimizer/__init__.py:167]: Since DeepSpeedCPUAdam also uses an offloaded grad, the gradient is stored in `p.grad` not `p.main_grad`. (regardless of DDP_impl)

File: megatron/optimizer/optimizer.py (hunk 10 above)
Comment (ttochi) [reply megatron/optimizer/optimizer.py:430]: 기존에 `MegatronOptimizer`는 grad_scaler를 가지는데, staged optimizer의 경우 grad_scaler는 `SpiralStageOptimizer`에서 global하게 관리하고 있어서 각 MegatronOptimizer가 자체적으로 업데이트하면 안됨.

이를 방지하기 위해 우선은 `self.use_global_grad_scaler` 라는 변수를 만들었는데, 이 부분은 다음 PR에서 삭제될 예정입니다. (mixed precision 로직을 inner optimizer로 옮길 예정이므로)

File: megatron/spiral/init_context.py (hunk 12 above)
Comment (ttochi) [reply megatron/spiral/init_context.py:473]: In `MegatronOptimizer.step`, gradient is freed during the optimizer process.

File: megatron/spiral/init_context.py (hunk 11 above)
Comment (ttochi) [reply megatron/spiral/init_context.py:377]: I'm worried that using a warning for every fetch will make the logs too noisy.😢

File: megatron/spiral/schedules/mobius_schedule.py (hunk 3 above)
Comment (ttochi) [reply megatron/spiral/schedules/mobius_schedule.py:563]: Same reason with above

File: megatron/optimizer/__init__.py (hunk 7 above)
Comment (ttochi) [reply megatron/optimizer/__init__.py:66]: stage optimizer를 사용하는 경우 gpu param을 optimizer의 param_group에 넘겨주게 되는데, 이 때 이미 offload된 gpu param은 빈 텐서라 spiral_shape을 사용하도록 수정함.

File: megatron/spiral/schedules/mobius_schedule.py (hunk 4 above)
Comment (gajagajago) [reply megatron/spiral/schedules/mobius_schedule.py]: good

File: megatron/optimizer/optimizer.py (hunk 10 above)
Comment (gajagajago) [reply megatron/optimizer/optimizer.py:430]: thx

File: megatron/spiral/init_context.py (hunk 11 above)
Comment (gajagajago) [reply megatron/spiral/init_context.py:377]: just notes plz then :0

**Comment by gajagajago** [comment]
@ttochi Let's merge this pr.

**Comment by ttochi** [comment]
Reflected the remaining reviews.
